SPEC.md

```diff
@@ -211,6 +211,65 @@ Plugins may define additional fields that they accept and may generate an error
 }
 ```
 
+### Network Configuration Lists
+
+Network configuration lists provide a mechanism to run multiple CNI plugins for a single container in a defined order, passing the result of each plugin to the next plugin.
```

Comment:
> In a world where Kubernetes has a Network object, it seems that object would translate to CNI plugins.

@thockin I would expect that a world with Kube Network objects would map whatever the kube network object was to a single CNI plugin config file, or a single CNI plugin config list.  Much like it does now, except that kube just picks the first "network" in the directory.

So, say Kube Network objects have a name.  You have some networks (a) "pod-network" and (b) "storage-network", (c) "really-cool-nfv-stuff".  You also have three CNI network config lists named /etc/cni/networks/pod-network.conflist, /etc/cni/networks/storage-network.conflist, and /etc/cni/networks/really-cool-nfv-stuff.conflist.

Most pods are tagged with "pod-network".  Only some pods (like your database) are tagged with "storage-network".  You have some firewall-as-a-service tagged with only "really-cool-nfv-stuff".

The storage-network conflist doesn't include things like hostport, since the database isn't serving any clients on the storage network, it's just consuming storage.

The really-cool-nfv-stuff pod uses some custom network plugin stuff, and so its conflist only has one plugin in it and none of the standard CNI ones.

>  Do we expect the user who creates the Network block to be forced to specify both portmapping and 'real" driver? That would force end-users to know about CNI. Or do we expect them to only specify the real driver, forcing kubelet or dockershim to add the portmapping plugin?

I don't think we can get away from this.  We have two choices: (a) make Kube Network objects a representation of some other object defined outside kube, or (b) force plugin writers and backends to translate IPAM, hostport, interface name, whatever details from a highly-specified Kube network object into something they can actually run.

I really don't think we want Kube to start carrying full network specifications, because there's so much variability in how plugins operate and what configuration they need.  eg, I don't see kube ever creating docker networks or CNI config files on its own, because we'll never get a standard network config format that works for most plugin vendors, and even if we did it would always be less capable.

Maybe I'm setting up a strawman here, but I think Kube Network objects should probably only hold basic information and essentially "refer" to some other network object that the network plugin actually knows about.  For CNI that would be a config file name or a conflist name.  For Docker that could be a docker network.

>  How does kubelet know that the driver specified by th euser doesn't do it natively?

Kubelet probably can't.  Depending on the network plugin the cluster has chosen, or the network plugin specified in the Network object, the user may have to do some small configuration.  Kubelet can provide all the information in a standardized way, but asking network plugins to optionally consume that information has two problems.  First, it creates plugins that are kubernetes-specific, which historically you did not want.  Second, it requires some system of capabilities of network plugins to ensure that if the user wants hostports but hte plugin doesn't support it, that an error occurs even if the plugin doesn't recognize hostports.